File: LC15_3Sum.py

"""
TC = O(n*n)
SC+O(n)

"""
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums.sort()
        for i in range(len(nums)):
            if nums[i]>0:
                break
            if i == 0 or nums[i-1]!=nums[i]:
                self.twoSum(nums, i ,res)
        return res
    def twoSum(self,nums, i, res):
        seen = set()
        j=i+1
        while j<len(nums):
            complement = -nums[i]-nums[j]
            if complement in seen:
                res.append([complement, nums[i], nums[j]])
                while j+1<len(nums) and nums[j]==nums[j+1]:
                    j+=1
            seen.add(nums[j])
            j+=1
                
"""
TC = O(n*n)
SC+O(n)
"""
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums.sort()
        for i in range(len(nums)):
            if nums[i]>0:
                break
            if i == 0 or nums[i-1]!=nums[i]:
                self.twoSumII(nums, i ,res)
        return res
    def twoSumII(self,nums, i, res):
        low = i+1
        high = len(nums)-1
        while(low<high):
            sum = nums[i]+nums[low]+nums[high]
            if sum<0:
                low+=1
            elif sum>0:
                high-=1
            else:
                res.append([nums[i], nums[low], nums[high]])
                low+=1
                high-=1
                while low<high and nums[low]==nums[low-1]:
                    low+=1
                    

File: LC15-3sum.java
import java.util.*;

class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        // brute force : find all possible pairs and check O(n^3)

        // for 2sum approach, using hashset we get Time O(n) and Space O(n)
        // for 2sum approach, using sort we get Time O(nlogn) and Space O(1)

        // for each, use 2sum approach to get target - cur, O(n * (O(2sum))) ans Space O(n)
        // Another idea, sort it and use 2 pointers, it will be O(n^2) and Space O(1)

        int n = nums.length;
        Arrays.sort(nums);

        List<List<Integer>> res = new ArrayList<>();

       for (int i=0; i<n-2; i++){
           // outsied 
           if (i>0 && nums[i]==nums[i-1]) continue;

           int l = i+1;
           int r = n-1;

           while(l<r) {
               int sum = nums[i] + nums[l] + nums[r];

               if (sum==0){
                   List<Integer> li = Arrays.asList(nums[i], nums[l], nums[r]);
                   res.add(li);
                   l++;
                   r--;
                   while (l<r && nums[l] == nums[l-1]) l++;
                   while (l<r && nums[r] == nums[r+1]) r--;
               } else if (sum >0){
                   r--;
               } else {
                   l++;
               }
           }
       }
       return res;
    }
}